backend/src/import.cpp

```diff
@@ -0,0 +1,311 @@
+#include <atomic>
+#include <condition_variable>
+#include <functional>
+#include <iostream>
+#include <memory>
+#include <mutex>
+#include <ranges>
+#include <span>
+#include <sstream>
+#include <string>
+#include <thread>
+#include <unordered_map>
+#include <vector>
+
+#include "sqlite3.h"
+
+using namespace std;
+
+constexpr auto BATCH_SIZE = 1024;         // records per insert statement
+constexpr auto CHUNK_SIZE = 1024 * 1024;  // size of stdin chunks to parse
+
+struct Node {
+    int idx;
+    string hash;
+    string data;
+    string type;
+};
+
+struct Edge {
+    int source;
+    int target;
+    int groupNum;
+};
+
+struct Graph {
+    vector<Node> nodes;
+    vector<Edge> edges;
+};
+
+enum class ParseResult {
+    DONE,
+    MORE,
+    FAIL,
+};
+
+class ParseState {
+    public:
+        ParseState() : buffer(1, '\0'), position(0) { }
+
+        void advance(int n) { position += n; }
+
+        auto remaining() {
+            auto pos = buffer.cbegin() + position;
+            return string_view(pos, buffer.cend());
+        }
+
+        bool feed(int n) {
+            buffer.erase(buffer.begin(), buffer.begin() + position);
+            position = 0;
+            auto readPosition = buffer.size() - 1;
+            buffer.resize(buffer.size() + n);
+            cin.read(&buffer.front() + readPosition, n);
+            auto actualRead = cin.gcount();
+            buffer.resize(readPosition + actualRead + 1);
+            buffer.back() = '\0';
+            return actualRead < n;
+        }
+
+        void error(const char* msg) {
+            cerr << "\x1b[31mparse error:\x1b[0m " << msg << ":\n";
+            cerr << remaining().substr(0, 256) << endl;
+        }
+
+    private:
+        vector<char> buffer;
+        int position;
+};
+
+ParseResult parseWarning(ParseState& state) {
+    if (!state.remaining().starts_with("Warning:")) {
+        state.error("no warning message");
+        return ParseResult::FAIL;
+    }
+    auto end = state.remaining().find("\n\n");
+    if (end == string_view::npos) {
+        state.error("expected blank line after warning");
+        return ParseResult::FAIL;
+    }
+    state.advance(end);
+    return ParseResult::DONE;
+}
+
+int assignNodeIdx(Node& node) {
+    static auto idxMap = unordered_map<size_t, int>();
+    static auto hashFn = hash<string>();
+    auto hash = hashFn(node.data);
+    auto [idx, _] = idxMap.emplace(make_pair(hash, idxMap.size() + 1));
+    node.hash = to_string(hash);
+    node.idx = idx->second;
+    return node.idx;
+}
+
+ParseResult parseNodeBazel5(ParseState& state, Graph& graph) { /*
+    BLUE_NODE:NodeData{e247f4c}|
+    RED_NODE:NodeData{6566162}|GREEN_NODE:NodeData{aab8e2e}|BLUE_NODE:NodeData{e247f4c}|RED_NODE:NodeData{5cedeee}
+    RED_NODE:NodeData{5cedeee}|GREEN_NODE:NodeData{aab8e2e}
+    GREEN_NODE:NodeData{aab8e2e}|RED_NODE:NodeData{5cedeee}
```

Comment:
Example of Skyframe dump format from Bazel 5. One call to `parseNodeBazel5` parses a single line of this.

Node keys are separated by `|` characters. The first node in a line is the source, the others are targets.
So `A|B|C|D` describes three edges: `A -> B`, `A -> C` and `A -> D`.